psbt/finalizer.go

```diff
@@ -0,0 +1,497 @@
+		positionMap = append(positionMap, kv{Key: pos, Value: p})
+	}
+
+	sort.Slice(positionMap, func(i, j int) bool {
+		return positionMap[i].Key < positionMap[j].Key
+	})
+	// Build the return array of signatures
+	sigsNew := [][]byte{}
+	for _, x := range positionMap {
+		sigsNew = append(sigsNew, x.Value[1])
+	}
+	return sigsNew, nil
+}
+
+// getMultisigScriptWitness creates a full Witness field for the transaction,
+// given the public keys and signatures to be appended, assuming (currently
+// without validating; parseScript from txscript is not exported so it can't
+// be checked here) that the witnessScript is of type M of N multisig. This
+// is used for both p2wsh and nested p2wsh multisig cases.
+func getMultisigScriptWitness(witnessScript []byte, pubKeys [][]byte,
+	sigs [][]byte) ([]byte, error) {
+
+	orderedSigs, err := extractKeyOrderFromScript(witnessScript, pubKeys, sigs)
+	if err != nil {
+		return nil, ErrInvalidPsbtFormat
+	}
+
+	var buf bytes.Buffer
+	var witnessItems = [][]byte{
+		nil}
+	for _, os := range orderedSigs {
+		witnessItems = append(witnessItems, os)
+	}
+	witnessItems = append(witnessItems, witnessScript)
+	err = writeTxWitness(&buf, witnessItems)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// checkSigHashFlags compares the sighash flag byte on a signature with the
+// value expected according to any PsbtInSighashType field in this section
+// of the PSBT, and returns true if they match, false otherwise.
+// If no SighashType field exists, it is assumed to be SIGHASH_ALL.
+// TODO taking the first byte of the field's value is not correct in general.
+// TODO Use txscript.SigHashAll
+func checkSigHashFlags(sig []byte, input *PsbtKeyValuePairList) bool {
+	expectedSighashType := byte(1)
+	if input.containsKeyType(PsbtInSighashType) {
+		sighashField, _ := input.getValueFromKeyType(PsbtInSighashType)
+		expectedSighashType = sighashField[0]
+	}
+	return expectedSighashType == sig[len(sig)-1]
+}
+
+// isFinalized considers this input finalized if it contains
+// at least one of the FinalScriptSig or FinalScriptWitness
+// are filled (which only occurs in a successful call to Finalize*)
+func isFinalized(p *Psbt, idx int) bool {
+	kvPairList := &(*p.InKVPairs)[idx]
+	if kvPairList.containsKeyType(PsbtInFinalScriptSig) ||
+		kvPairList.containsKeyType(PsbtInFinalScriptWitness) {
+		return true
+	}
+	return false
+}
+
+// isFinalizable checks whether the structure of the entry
+// for the input of the Psbt p at index idx contains sufficient
+// information to finalize this input, based on the template
+// implied by addrType (currently only 'p2wpkh, 'p2wsh', 'np2wpkh' or
+// 'np2wsh' are accepted, where 'n' refers to nested).
+// TODO Use something from the lib (not string) for address type,
+// and implement the code for other address types.
+func isFinalizable(p *Psbt, idx int, addrType string) bool {
+	iKVPLs := *(p.InKVPairs)
+	iKVPL := iKVPLs[idx]
+	switch addrType {
+	case "p2wpkh":
```

Comment:
The named script-hash script types feels wrong.. what about passing around (isP2sh, isP2wsh, script)? Then the code you write is more generalizable towards [, P2SH, P2WSH, P2SH|P2WSH] handling. (you don't want to write multisig support for each script-hash type or for each m-of-n for that matter)

Reply:
Yes, as noted in the comment just above, this is definitely inappropriate and should be replaced with something in-lib.